torchao/prototype/float8nocompile/float8nocompile_linear.py

```diff
@@ -145,3 +109,94 @@ def from_float(cls, mod, kernel_algo: KernelAlgorithm = KernelAlgorithm.ATOMIC_M
 
         # TODO(danielvegamyhre): support for FSDP once dependencies are implemented
         return new_mod
+
+
+class matmul_with_args_in_hp(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, input_hp, weight_hp, config, linear_mm_config, kernel_algo):
+        # output = input @ weight_t
+        input_fp8_row_major, input_fp8_col_major = ToFP8RowAndColumnMajor.apply(
+            input_hp,
+            config.cast_config_input.target_dtype,
+            linear_mm_config,
+            GemmInputRole.INPUT,
+            kernel_algo,
+        )
+        weight_t_fp8_col_major = ToFP8ColumnMajorT.apply(
+            weight_hp,
+            config.cast_config_weight.target_dtype,
+            linear_mm_config,
+            GemmInputRole.WEIGHT,
+            kernel_algo,
+        )
+        output = torch.mm(input_fp8_row_major, weight_t_fp8_col_major)
+
+        # save data for backward before returning
+        ctx.save_for_backward(input_fp8_col_major, weight_hp)
+        ctx.config = config
+        ctx.linear_mm_config = linear_mm_config
+        ctx.kernel_algo = kernel_algo
+
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        input_fp8_col_major, weight_hp = ctx.saved_tensors
+
+        # cast grad output to float8_e5m2 for backward
+        grad_output_fp8_row_major = ToFP8RowMajor.apply(
+            grad_output,
+            ctx.config.cast_config_grad_output.target_dtype,
+            ctx.linear_mm_config,
+            GemmInputRole.GRAD_OUTPUT,
+            ctx.kernel_algo,
+        )
+
+        # grad_input = grad_output @ weight
+        weight_fp8_col_major = ToFP8ColumnMajor.apply(
+            weight_hp,
+            ctx.config.cast_config_weight.target_dtype,
+            ctx.linear_mm_config,
+            GemmInputRole.WEIGHT,
+            ctx.kernel_algo,
+        )
+        grad_input = torch.mm(grad_output_fp8_row_major, weight_fp8_col_major)
+
+        # grad_weight = grad_output_t @ input
+        # apparently this variant is slightly faster than `grad_weight_t = input_t @ grad_output`
+        # source: https://github.com/pytorch/ao/blob/fe5f11b2c58b452e01ba9ec7359629928b143619/torchao/float8/float8_linear.py#L84-L85
+        grad_output_t_row_major = ToFP8RowMajorT.apply(
+            grad_output,
+            ctx.config.cast_config_grad_output.target_dtype,
+            ctx.linear_mm_config,
+            GemmInputRole.GRAD_OUTPUT,
+            ctx.kernel_algo,
+        )
+        grad_weight = torch.mm(grad_output_t_row_major, input_fp8_col_major)
+        return grad_input, grad_weight, None, None, None
+
+
+class matmul_with_args_in_fp8(torch.autograd.Function):
```

Comment:
this function should assert that the memory format is expected for all arguments

Reply:
Done